[Feat] 건물 또는 편의시설 검색 기능 구현

lyouxsun opened the pull request:
## Related issue 🛠
- closed #23 

## Work Description 📝
- 건물 또는 편의시설 검색 기능을 구현했습니다.
- 사용자가 검색어를 입력하면, 먼저 해당 검색어가 건물 이름에 포함되어 있는지를 확인합니다.
    - 검색어에 해당하는 건물이 존재할 경우, 해당 건물 정보와 그 건물에 속한 모든 편의시설 목록을 함께 반환합니다.
    - 검색어에 해당하는 건물이 없을 경우, 검색어를 포함하는 편의시설 목록만 조회하여 반환합니다.
- 응답 형태는 두 가지로 구분되며, type 필드를 통해 결과 종류를 명시합니다.
    - "BUILDING" : 건물 + 해당 건물에 포함된 편의시설 목록
    - "FACILITY" : 검색어에 해당하는 편의시설 목록만 반환

## Screenshot 📸
- type이 "BUILDING" 인 경우
    <img width="300" alt="스크린샷 2025-07-27 오후 9 20 07" src="https://github.com/user-attachments/assets/5b646b74-ca77-4d43-af3b-308978e383d8" />

- type이 "FACILITY" 인 경우
    <img width="300" alt="스크린샷 2025-07-27 오후 9 20 42" src="https://github.com/user-attachments/assets/e9e8aea0-1c65-4d72-9bca-9e01b1d03167" />


## Uncompleted Tasks 😅
- [ ] 사용자의 현재 위치를 쿼리 파라미터로 받아서 직선 거리 순으로 오름차순 정렬을 해서 보내주기 위해 응답에 위도와 경도 데이터가 들어있습니다. 확장성을 생각하여 추가해두긴 했지만 아직은 구현하지 못한 상태입니다. 😅

## To Reviewers 📢
응답을 위한 DTO 종류가 많아 혼란스러울 수 있습니다.
- 모든 응답은 `PlaceSearchResponse` 형식으로 통일되어 있으며, type 값에 따라 data 필드에 담기는 객체가 달라집니다.
    - type이 "BUILDING"인 경우 → data에는 SearchByBuildingNameResponse 객체가 들어갑니다.
    - type이 "FACILITY"인 경우 → data에는 SearchByFacilityNameResponse 객체가 들어갑니다.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **신규 기능**
  * 장소 검색을 위한 REST API 엔드포인트가 추가되었습니다. 키워드로 건물 및 시설을 검색할 수 있습니다.
  * 건물 및 시설 검색 결과를 반환하는 응답 구조가 추가되었습니다.
  * 건물명 및 시설명을 기준으로 부분 일치(대소문자 구분 없이) 검색이 가능합니다.

* **리팩터링**
  * 일부 DTO 클래스가 Java record 문법으로 단순화되어 불변성과 코드 가독성이 향상되었습니다.

* **환경 설정**
  * 로컬 데이터베이스의 스키마 생성 전략이 매번 새로 생성(create)으로 변경되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lyouxsun committed:
[Feat] #23 장소 검색 API 구현

- 키워드 기반으로 장소(건물/편의시설) 검색 가능
- 건물명이 포함된 경우: 해당 건물과 편의시설 목록 반환
- 건물명이 없는 경우: 편의시설 기준으로 건물 정보 포함 응답

lyouxsun committed:
[Docs] #23 Swagger 문서화를 위한 어노테이션 추가

lyouxsun committed:
[Fix] #23 건물 검색 시 응답 dto 수정

yeonzeop commented on src/main/java/com/example/BarrierKU/domain/building/dto/BuildingResponse.java:
```diff
@@ -98,7 +98,7 @@ public BuildingResponse(Long buildingId, int number, String name, String departm
         this.facilityPurposes = purposes;
-        this.doorInfos = doors.stream().map(door -> new DoorInfo(door)).toList();
-        this.significantInfos = significants.stream().map(significant -> new SignificantInfo(significant)).toList();
+        this.doorInfos = doors.stream().map(DoorInfo::new).toList();
```
감사합니당 헤헤

yeonzeop commented on src/main/java/com/example/BarrierKU/domain/place/service/PlaceService.java:
```diff
@@ -0,0 +1,40 @@
+        if (!buildings.isEmpty()) {
+            // 1. 만약 건물이 하나라도 조회된다.
+            log.debug("[getPlaces] 검색어를 포함한 건물이름 존재 = {}",
+                    buildings.stream().map(SearchBuildingResponse::name).collect(Collectors.joining(", ")));
```
수고 많으셨어요!! 응답 구조가 단일하지 않아서 그런지, 읽기만하는 제 입장에서도 참 어렵네요 ㅜㅜ 멋있으십니다 //_//

yeonzeop reviewed (approved):
수고많으셨습니다! Repository에서 사용하신 메소드들도 그렇고 처음 보는 게 많네요!! 열심히 공부해보겠습니다 🤓🤓😍😍

lyouxsun committed:
[Fix] #23 merge conflict 해결

ksg1227 reviewed (approved):
수고하셨습니다! 제네릭을 활용해서 코드를 깔끔하게 잘 짜주신 것 같습니다.
여기서 두 가지정도 궁금증이 있습니다!

1. 현재 구현처럼 하나의 API 에서 와일드카드를 통해 여러 개의 타입을 커버할 수 있도록 구현할 경우, 스웨거에 명세는 어떤 식으로 표현되나요? 

2. 현재 구현 방식처럼, 서버 단에서 알아서 검색된 장소의 타입을 구분하여 응답을 분리해서 내려주는 방식도 가능할 것 같고, 검색어를 먼저 서버로 전송하면, 해당 검색어가 BUILDING 인지, FACILITY 인지 클라이언트에게 알려주고, 클라이언트가 그 타입에 따라 서로 다른 API 로 요청을 분기해서 반환하는 방식도 있을 것 같다는 생각이 듭니다. 응답 구조가 다르다보니까 그러한 생각이 드네요!

다만 현재와 같이 하나의 API 에서 다양한 형태의 응답을 반환하는 경우, 클라이언트 측에서 이를 처리하기에 구현 난이도가 어느정도일지 잘 가늠이 안가네요.. 그래서 혹시 이부분에 대한 논의가 클라쪽과 이루어졌는지 궁금합니다

lyouxsun commented:
1. `@Schema(oneOf = {SearchByBuildingNameResponse.class, SearchByFacilityNameResponse.class})` 어노테이션을 명시적으로 사용하면 type 값에 따라 스웨거에서 응답 예시를 분기할 수 있다고 하네요! 적용해 두겠습니다.

2. 사용자가 검색어를 입력할 때 한 음절이 완성될 때마다 서버에게 요청을 보내기 때문에, API를 분리해서 처리하기보다는 한번의 요청으로 통합 처리하는 것이 성능 측면에서 효율적일 것이라고 판단했습니다. 응답 형태가 다르더라도 type 변수값을 통해 프론트가 분기 처리가 수월할 것이라고 생각했지만, 이 부분은 프론트 개발자들과 이야기를 더 해봐야 할 것 같습니다!

lyouxsun committed:
[Fix] #23 swagger 명세 수정

ksg1227 commented:
> 1. `@Schema(oneOf = {SearchByBuildingNameResponse.class, SearchByFacilityNameResponse.class})` 어노테이션을 명시적으로 사용하면 type 값에 따라 스웨거에서 응답 예시를 분기할 수 있다고 하네요! 적용해 두겠습니다.
> 2. 사용자가 검색어를 입력할 때 한 음절이 완성될 때마다 서버에게 요청을 보내기 때문에, API를 분리해서 처리하기보다는 한번의 요청으로 통합 처리하는 것이 성능 측면에서 효율적일 것이라고 판단했습니다. 응답 형태가 다르더라도 type 변수값을 통해 프론트가 분기 처리가 수월할 것이라고 생각했지만, 이 부분은 프론트 개발자들과 이야기를 더 해봐야 할 것 같습니다!

아하 한 음절마다 요청을 보내는 형태군요. 확실히 성능이 아주 중요한 부분이지 않을까 생각이 드네요.. 영선님 말씀처럼 하나의 API 로 관리하는 것이 요청을 한 번으로 줄일 수 있는 방법이지 않을까 생각이 드는데, 그럼 우선은 프론트분들과 이야기 완료되면 원하는 방식에 맞춰주면 될 것 같습니다! 
추후에는 여기에는 Redis 를 활용해서 캐싱을 하는 것도 좋을 것 같다는 생각도 드네요

choihongjun1 commented on src/main/java/com/example/BarrierKU/domain/place/controller/PlaceController.java:
```diff
@@ -0,0 +1,48 @@
+                    @Parameter(name = "keyword", description = "검색할 키워드 (건물명 또는 편의시설명)", required = true)
+            }
+    )
+    @CustomExceptionDescription(GET_SPACE_INFO)
```
검색 결과가 없을 때 예외가 발생하는 건가요?
저는 검색결과가 없을 때 빈리스트를 반환하고 성공으로 했는데 수정해야 하는지 궁금합니다!

GET_SPACE_INFO는 room 조회할 때 발생하게 추가해둔건데, 여기랑 구분 안해도 될까요?

choihongjun1 commented on src/main/java/com/example/BarrierKU/domain/facility/dto/FacilitiesResponse.java:
```diff
@@ -1,36 +1,21 @@
-        private double longitude;
-    }
-
+public record FacilitiesResponse(
```
swagger 문서화 안되어 있습니다!

choihongjun1 commented on src/main/java/com/example/BarrierKU/domain/place/dto/SearchFacilityResponse.java:
```diff
@@ -0,0 +1,28 @@
+    private String name;
+
+    @Schema(description = "시설 용도", example = "CAFE")
+    private Purpose purpose;
```
건물 페이지의 주요시설에서는 purpose에 따라 아이콘+용도를 보여주기 때문에 purpose의 value를 반환하는데, 여기서는 아이콘만 보여주기 때문에 enum을 반환하는 것 같습니다.
프론트가 purpose 값에 따라서 아이콘을 가져다 쓰는거면, value로 왔을때랑 enum으로 왔을때 둘 다 처리해야 돼서 비효율적으로 되는게 아닌지 궁금합니다.

choihongjun1 reviewed (commented):
수고하셨습니다! 좀 헷갈려서 이해하는데 오래 걸렸네요..

lyouxsun committed:
refactor: 건물 정보 유무와 관계없이 응답 형태를 일관되게 반환하도록 개선

- 건물 검색 결과가 없을 경우에도 buildings, facilities 필드를 빈 리스트로 유지
- 클라이언트가 응답 구조를 예측 가능하도록 일관성 확보

lyouxsun committed:
Merge remote-tracking branch 'origin/develop' into feat/#23-search-place
